source/common/ssl/context_impl.cc

```diff
@@ -365,11 +379,120 @@ ServerContextImpl::ServerContextImpl(ContextManagerImpl& parent, Stats::Scope& s
+    }
+    SSL_CTX_set_tlsext_ticket_key_cb(ctx_.get(), [](SSL* ssl, uint8_t* key_name, uint8_t* iv,
+                                                    EVP_CIPHER_CTX* ctx, HMAC_CTX* hmac_ctx,
+                                                    int encrypt) -> int {
+      ContextImpl* context_impl =
+          static_cast<ContextImpl*>(SSL_CTX_get_ex_data(SSL_get_SSL_CTX(ssl), ssl_context_index_));
+      return dynamic_cast<ServerContextImpl*>(context_impl)
+          ->sessionTicketProcess(ssl, key_name, iv, ctx, hmac_ctx, encrypt);
+    });
+  }
+
+  uint8_t session_context_buf[EVP_MAX_MD_SIZE] = {};
+  unsigned session_context_len = 0;
+  if (ca_cert_ != nullptr) {
+    // If there is a CA cert, only allow resuming in a context with the same
+    // CA cert.  This ensures that the client is always validated against
+    // the correct CA cert, even if session resumption across different listeners
+    // is enabled.
+    int ret = X509_digest(ca_cert_.get(), EVP_sha256(), session_context_buf, &session_context_len);
+    if (ret != 1) {
+      throw EnvoyException("Failed to hash ca_cert");
+    }
+  } else {
+    memcpy(session_context_buf, &SERVER_SESSION_ID_CONTEXT, sizeof(SERVER_SESSION_ID_CONTEXT));
+    session_context_len = sizeof(SERVER_SESSION_ID_CONTEXT);
+  }
+
+  int ret = SSL_CTX_set_session_id_context(ctx_.get(), session_context_buf, session_context_len);
+  if (ret != 1) {
+    throw EnvoyException("Failed to set session id context");
+  }
+}
+
+int ServerContextImpl::sessionTicketProcess(SSL*, uint8_t* key_name, uint8_t* iv,
+                                            EVP_CIPHER_CTX* ctx, HMAC_CTX* hmac_ctx, int encrypt) {
+  const EVP_MD* hmac = EVP_sha256();
+  const EVP_CIPHER* cipher = EVP_aes_256_cbc();
+
+  if (encrypt == 1) {
+    // Encrypt
+    RELEASE_ASSERT(session_ticket_keys_.size() >= 1);
+    const SessionTicketKey& enc_key = session_ticket_keys_.front();
+
+    RELEASE_ASSERT(enc_key.name.size() == SSL_TICKET_KEY_NAME_LEN);
+    std::copy_n(enc_key.name.begin(), SSL_TICKET_KEY_NAME_LEN, key_name);
+
+    if (!RAND_bytes(iv, EVP_CIPHER_iv_length(cipher))) {
+      return -1;
+    }
+
+    ASSERT(enc_key.aes_key.size() == EVP_CIPHER_key_length(cipher));
+    if (!EVP_EncryptInit_ex(ctx, cipher, nullptr, enc_key.aes_key.data(), iv)) {
+      return -1;
+    }
+
+    if (!HMAC_Init_ex(hmac_ctx, enc_key.hmac_key.data(), enc_key.hmac_key.size(), hmac, nullptr)) {
+      return -1;
+    }
+
+    return 1; // success
+  } else {
+    // Decrypt
+    bool is_enc_key = true; // first element is the encryption key
+    for (const SessionTicketKey& key : session_ticket_keys_) {
+      RELEASE_ASSERT(key.name.size() == SSL_TICKET_KEY_NAME_LEN);
+      if (std::equal(key.name.begin(), key.name.end(), key_name)) {
+        if (!HMAC_Init_ex(hmac_ctx, key.hmac_key.data(), key.hmac_key.size(), hmac, nullptr)) {
+          return 0;
+        }
+
+        ASSERT(key.aes_key.size() == EVP_CIPHER_key_length(cipher));
```

Comment:
This should be checked when adding key to the `session_ticket_keys_` and not when using it.